src/node_large_page.cc

```diff
@@ -0,0 +1,233 @@
+// Copyright (C) 2018 Intel Corporation
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom
+// the Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+// OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+// OR OTHER DEALINGS IN THE SOFTWARE.
+//
+// SPDX-License-Identifier: MIT
+
+#include <errno.h>
+#include <fcntl.h>  // _O_RDWR
+#include <limits.h>  // PATH_MAX
+#include <locale.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <sys/mman.h>
+
+#include <string>
+#include <fstream>
+#include <iostream>
+
+// The functions in this file map the text segment of node into 2M pages.
+// The algorithm is quite simple
+// 1. Find the text region of node binary in memory
+// 2. Move the text region to large pages
+
+extern char __executable_start;
+extern char __etext;
+extern char __nodetext;
+
+namespace node {
+namespace largepages {
+#define ALIGN(x, a)           (((x) + (a) - 1) & ~((a) - 1))
+#define PAGE_ALIGN_UP(x, a)   ALIGN(x, a)
+#define PAGE_ALIGN_DOWN(x, a) ((x) & ~((a) - 1))
+
+struct TextRegion {
+  void * from;
+  void * to;
+  int    totalHugePages;
+  int64_t  offset;
+  bool   found_text_region;
+};
+
+    static void printSystemError(int error) {
+      fprintf(stderr, "Hugepages WARNING: %s\n", strerror(error));
+      return;
+    }
+
+    static struct TextRegion find_node_text_region() {
+      FILE *f;
+      int64_t  start, end, offset, inode;
+      char perm[5], dev[6], name[256];
+      int ret;
+      const size_t hugePageSize = 2L * 1024 * 1024;
```

Comment:
does it have a bearing to existing Linux attributes? Are they same across releases and vendors? are they tunable at OS level?

Reply:
The structure of /proc/<pid>/maps is consistent across many different releases/vendors.

man proc
       /proc/[pid]/maps
              A file containing the currently mapped memory regions and their access  permissions.   

              The format of the file is:
       address           perms offset  dev   inode       pathname
       00400000-00452000 r-xp 00000000 08:02 173521      /usr/bin/dbus-daemon
       00651000-00652000 r--p 00051000 08:02 173521      /usr/bin/dbus-daemon
       00652000-00655000 rw-p 00052000 08:02 173521      /usr/bin/dbus-daemon
       00e03000-00e24000 rw-p 00000000 00:00 0           [heap]
       00e24000-011f7000 rw-p 00000000 00:00 0           [heap]
       ...
       35b1800000-35b1820000 r-xp 00000000 08:02 135522  /usr/lib64/ld-2.15.so
       35b1a1f000-35b1a20000 r--p 0001f000 08:02 135522  /usr/lib64/ld-2.15.so
       35b1a20000-35b1a21000 rw-p 00020000 08:02 135522  /usr/lib64/ld-2.15.so
       35b1a21000-35b1a22000 rw-p 00000000 00:00 0

For the node process it looks like this.
00400000-02168000 r-xp 00000000 08:01 541586     /home/dslo/node-ssuresh-fork/node
02367000-02369000 r--p 01d67000 08:01 541586      /home/dslo/node-ssuresh-fork/node
02369000-02380000 rw-p 01d69000 08:01 541586     /home/dslo/node-ssuresh-fork/node
02380000-0248f000 rw-p 00000000 00:00 0                                  [heap]

After we map part of the area, you will see the text spilt into 3 executable pages. I will put this in the code comment as well.

00400000-00a00000 r-xp 00000000 08:01 541593    /home/dslo/node-ssuresh-fork/node.large
00a00000-02200000 r-xp 00000000 00:00 0 
02200000-02397000 r-xp 01e00000 08:01 541593    /home/dslo/node-ssuresh-fork/node.large
02597000-02598000 r--p 01f97000 08:01 541593     /home/dslo/node-ssuresh-fork/node.large
02598000-025af000 rw-p 01f98000 08:01 541593     /home/dslo/node-ssuresh-fork/node.large
025af000-025c7000 rw-p 00000000 00:00 0 
036df000-037d2000 rw-p 00000000 00:00 0                                  [heap]